edition = "2023";
option features.field_presence = IMPLICIT;
package semlog.core_engine;

service CoreEngineService {
    rpc PostIngestion(IngestLogRequest) returns (IngestLogResponse);
    rpc GetStatus(GetStatsRequest) returns (GetStatsResponse);
}

message IngestLogRequest {
    string service_name = 1;
    string timestamp = 2;
    string message = 3;
}

message IngestLogResponse {
    string action = 1;
    string semantic_id = 2;
    string compression_ratio = 3;
}

message PatternStats {
    string pattern = 1;
    int32 occurrence_count = 2;
}

message GetStatsRequest{
}

message GetStatsResponse {
    int32 unique_semantic_patterns = 1;
    int32 total_logs_absorbed = 2;
    string compression_rate = 3;
    repeated PatternStats top_patterns = 4;
}